File: POLY/returns_poly.cpp
//
//  main.cpp
//  Poly
//
//  Created by DooSeon Choi on 2014. 8. 24..
//  Copyright (c) 2014³â DoO. All rights reserved.
//

#include <iostream>

long gPolyMatrix[100][100];

long getPolyMatrixValue( int aColulmn, int aTarget )
{
    int sColumnIndex = 0;
    long sValue = 0;
    
    for ( sColumnIndex = 1; sColumnIndex <= aTarget - aColulmn; sColumnIndex++ )
    {
        sValue += gPolyMatrix[aTarget - aColulmn][sColumnIndex] * ( aColulmn + sColumnIndex - 1 );
    }
    
    return sValue % 10000000;
}

long getPolyValue( int aValue )
{
    long sValue = 0;
    int sIndex = 0;
    
    for ( sIndex = 1; sIndex < aValue + 1; sIndex++ )
    {
        sValue += gPolyMatrix[aValue][sIndex];
    }
    
    return sValue % 10000000;
}

void makePolyMatrix( int aCount )
{
    int sRowIndex = 0;
    int sColulmnIndex = 0;
    long sValue = 0;
    
    for ( sRowIndex = 1; sRowIndex < aCount + 1; sRowIndex++ )
    {
        for ( sColulmnIndex = 1; sColulmnIndex < sRowIndex + 1; sColulmnIndex++ )
        {
            if ( sRowIndex == sColulmnIndex )
            {
                gPolyMatrix[sRowIndex][sColulmnIndex] = 1;
            }
            else
            {
                sValue = getPolyMatrixValue( sColulmnIndex, sRowIndex );
                gPolyMatrix[sRowIndex][sColulmnIndex] = sValue;
            }
        }
        
        gPolyMatrix[sRowIndex][0] = getPolyValue( sRowIndex );
    }
}

int main(int argc, const char * argv[])
{
    int sCount = 0;
    int sValue = 0;
    int sIndex = 0;
    
    scanf("%d", &sCount );
    
    makePolyMatrix( 100 );
    
    for ( sIndex = 0; sIndex < sCount; sIndex ++ )
    {
        scanf("%d", &sValue );
        /*
        if ( gPolyMatrix[sValue][0] != 0 )
        {
        }
        else
        {
            makePolyMatrix( sValue );
        }
         */
        
        printf("%ld\n", gPolyMatrix[sValue][0] );
    }
    
    return 0;
}

File: POLY/returns_poly.py

gPolyMatrix = [ [ int(0) for i in range(101) ] for i in range(101) ]


def getPolyMatrixValue( aColumn, aTarget ) :

    global gPolyMatrix
    sValue = 0

    for sColumnIndex in range( 1, aTarget - aColumn + 1 ) :
        sTempValue = gPolyMatrix[aTarget-aColumn][sColumnIndex] * ( aColumn + sColumnIndex - 1 )
        sValue = sValue + sTempValue

    return sValue % 10000000

def getPolyValue( aValue ) :

    global gPolyMatrix
    sValue = 0

    for i in range( 1, aValue + 1 ):
        sValue += gPolyMatrix[aValue][i]

    return sValue % 10000000

def getPolyMatrix( aCount ) :

    global gPolyMatrix
    sValue = 0

    for sRowIndex in range( 1, aCount + 1 ) :
        for sColumnIndex in range( 1, sRowIndex + 1 ) :
            if sRowIndex is sColumnIndex :
                sValue = 1
            else :
                sValue = getPolyMatrixValue( sColumnIndex, sRowIndex )

            gPolyMatrix[sRowIndex][sColumnIndex] = sValue

        gPolyMatrix[sRowIndex][0] = getPolyValue( sRowIndex )

def solve() :

    global gPolyMatrix

    sCount = input()

    for i in range( sCount ) :
        sValue = input()

        if gPolyMatrix[sValue][0] is 0 :
            getPolyMatrix( sValue )

        print gPolyMatrix[sValue][0]

solve()
